src/devices/Uln2003/Uln2003.cs

```diff
@@ -0,0 +1,169 @@
+                 { false, false, true, true, false, false, true, true }
+            };
+
+        /// <summary>
+        /// Initialize a Uln2003 class.
+        /// </summary>
+        /// <param name="pin1">The GPIO pin number which corresponds pin A on ULN2003 driver board.</param>
+        /// <param name="pin2">The GPIO pin number which corresponds pin B on ULN2003 driver board.</param>
+        /// <param name="pin3">The GPIO pin number which corresponds pin C on ULN2003 driver board.</param>
+        /// <param name="pin4">The GPIO pin number which corresponds pin D on ULN2003 driver board.</param>
+        public Uln2003(int pin1, int pin2, int pin3, int pin4)
+        {
+            _pin1 = pin1;
+            _pin2 = pin2;
+            _pin3 = pin3;
+            _pin4 = pin4;
+
+            _controller = new GpioController();
+
+            _controller.OpenPin(_pin1, PinMode.Output);
+            _controller.OpenPin(_pin2, PinMode.Output);
+            _controller.OpenPin(_pin3, PinMode.Output);
+            _controller.OpenPin(_pin4, PinMode.Output);
+        }
+
+        /// <summary>
+        /// Sets the motor speed to revolutions per minute.
+        /// </summary>
+        /// <remarks>Default revolutions per minute for 28BYJ-48 is approximately 15.</remarks>
+        public short RPM { get; set; }
+
+        /// <summary>
+        /// Sets the stepper's mode.
+        /// </summary>
+        public StepperMode Mode
+        {
+            get
+            {
+                return _mode;
+            }
+            set
+            {
+                _mode = value;
+
+                switch (_mode)
+                {
+                    case StepperMode.HalfStep:
+                        _currentSwitchingSequence = _halfStepSequence;
+                        break;
+                    case StepperMode.FullStepSinglePhase:
+                        _currentSwitchingSequence = _fullStepSinglePhaseSequence;
+                        break;
+                    case StepperMode.FullStepDualPhase:
+                        _currentSwitchingSequence = _fullStepDualPhaseSequence;
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Stop the motor.
+        /// </summary>
+        public void Stop()
+        {
+            _steps = 0;
+            _stopwatch.Stop();
+            _controller.Write(_pin1, PinValue.Low);
+            _controller.Write(_pin2, PinValue.Low);
+            _controller.Write(_pin3, PinValue.Low);
+            _controller.Write(_pin4, PinValue.Low);
+        }
+
+        /// <summary>
+        /// Moves the motor. If the number is negative, the motor moves in the reverse direction.
+        /// </summary>
+        /// <param name="steps">Number of steps.</param>
+        public void Step(int steps)
+        {
+            double lastStepTime = 0;
+            _stopwatch = Stopwatch.StartNew();
```

Comment:
I'm not sure if we should care about that scenario, but I don't think that this is thread safe as you could have two threads calling step which would override the stopwatch. Also, it kind of feels that this should be an Async method.

Reply:
we do not guarantee thread safety for a single instance unless the doc says otherwise but I agree we should not allocate stopwatch here